### GymBeam interview exercise

##### The aim of this interview exercise is to estimate when the long_term_unaivailable and simple products will run out of stock.  

First I need to import all the for this task necessary Python libraries 

In [1]:
import numpy as np
import pandas as pd

Afterwards I will read the data from all the provided tables and display first 10 rows to have a quick look at the data each table and which data could be used for the analysis

In [2]:
# read the data from the .csv and display first 10 rows
stock_days_data = pd.read_csv('in.c_assignment_hash.cm_stock_days_data.csv.gz', compression='gzip', sep=',')
stock_days_data.head(10)

id      period                   product_id_hash  is_in_stock    qty  \
0  9992375  2019-02-23  936a0bd1ac3c96d08aa66bfbc3ed9758            0    0.0   
1  9992268  2019-02-23  7ca595cd7955654b36629c1470763487            1  297.0   
2  9989948  2019-02-23  07b1c04a30f798b5506c1ec5acfb9031            0    0.0   
3  9986736  2019-02-23  4ea83d951990d8bf07a68ec3e50f9156            1   56.0   
4  9987935  2019-02-23  b49d4455d64520060ac01fb5a3c757e4            0    0.0   
5  9986209  2019-02-23  1e48c4420b7073bc11916c6c1de226bb            0    0.0   
6  9991620  2019-02-23  53d7f154d6c0738fa10f9402b2e93e96            0    0.0   
7  9989187  2019-02-23  d8567273b20e64233b575a130159a15d            0    0.0   
8  9990412  2019-02-23  0912d0f15f1394268c66639e39b26215            0    0.0   
9  9989773  2019-02-23  604616e4d592b744e14ec3ff33204dec            0    0.0   

      cost  
0   3.0200  
1   1.2744  
2  12.8800  
3  19.6600  
4   4.2400  
5   8.5400  
6  11.4500  
7   5.0000  
8   0.0000  
9   2.3200

From stock_days_data table I could use the following columns: *product_id_hash*, *period*, *is_in_stock*, *qty*

In [3]:
stock_movement = pd.read_csv('in.c_assignment_hash.cm_stock_movement.csv.gz', compression='gzip', sep=',')
stock_movement.head(10)

movement_id                    entity_id_hash  order_id increment_id  \
0      2691340  9f3a4ff872f93c8de65efc2f2360ce77         0    100883036   
1      2689585  6915849303a3fe93657587cb9c469f00         0    100882862   
2      2688976  7a5400eeb415675960d6822b6bdffb7a         0    300627454   
3      2690326  a92df7cbfdddcc938abc806992026a19         0    300627502   
4      2691214  2e855f9489df0712b4bd8ea9e2848c5a         0    200667465   
5      2688517  c5f03386978cf62ddb7f2f46e9bd5790         0    400571498   
6      2688508  c5f03386978cf62ddb7f2f46e9bd5790         0    400571495   
7      2689192  e02721e864b2649003bcf15ba4da931a         0    200667312   
8      2689990  3090f54f8d8eecd6469c3a9eb3ddb48a         0    200667369   
9      2691217  caa202034f268232c26fac9435f54e15         0    100883021   

   customer_id                customer_name_hash  is_admin mvt_qty_type  \
0     129313.0  1f1447abc5999be3ca06ef555cdb5d83         0          dwn   
1          0.0                               NaN         0          dwn   
2      69352.0  3ae09bc452b171003c0a97a0775108f0         0          dwn   
3      33921.0  e1908e8e7e5a6b0ca21555014eb451ec         0          dwn   
4          0.0                               NaN         0          dwn   
5          0.0                               NaN         0          dwn   
6          0.0                               NaN         0          dwn   
7          0.0                               NaN         0          dwn   
8     122869.0  70bb7027f4fa0c4f4399fa88996c89e8         0          dwn   
9      13611.0  41738a7e33c3da4eeb7d0d0d43a4af41         0          dwn   

   mvt_qty     qty     cost stock_operation  \
0        1   148.0   1.9800    order_create   
1        1  6029.0   0.7400    order_create   
2        1   181.0   6.5400    order_create   
3        1  1288.0   2.7500    order_create   
4        1   498.0  15.6612    order_create   
5        1   598.0   3.1000    order_create   
6        1   598.0   3.1000    order_create   
7        1  1239.0   5.9100    order_create   
8        1   338.0   4.5600    order_create   
9        1  3128.0   8.1000    order_create   

                              comment  note               created_at  
0  Product ordered (order: 100883036)   NaN  2019-04-13 13:24:06.000  
1  Product ordered (order: 100882862)   NaN  2019-04-13 11:03:27.000  
2  Product ordered (order: 300627454)   NaN  2019-04-13 10:29:40.000  
3  Product ordered (order: 300627502)   NaN  2019-04-13 11:59:18.000  
4  Product ordered (order: 200667465)   NaN  2019-04-13 13:12:12.000  
5  Product ordered (order: 400571498)   NaN  2019-04-13 09:48:54.000  
6  Product ordered (order: 400571495)   NaN  2019-04-13 09:46:43.000  
7  Product ordered (order: 200667312)   NaN  2019-04-13 10:34:17.000  
8  Product ordered (order: 200667369)   NaN  2019-04-13 11:27:39.000  
9  Product ordered (order: 100883021)   NaN  2019-04-13 13:13:25.000

From stock_movement table I could use the following columns: *entity_id_hash*, *mvt_quantity_type*, *mvt_qty*, *qty*, *created_at*

In [4]:
stock_products = pd.read_csv('in.c_assignment_hash.cm_stock_products.csv.gz', compression='gzip', sep=',')
stock_products.head(10)

entity_id               created_at                update_at  updated_by  \
0      35146  2019-04-19 07:30:13.000  2019-04-19 07:30:13.000         NaN   
1      35140  2019-04-19 07:28:28.000  2019-04-19 07:28:28.000         NaN   
2      35158  2019-04-19 08:33:18.000  2019-04-19 08:33:18.000         NaN   
3      35155  2019-04-19 08:31:00.000  2019-04-19 08:31:00.000         NaN   
4      35143  2019-04-19 07:28:28.000  2019-04-19 07:28:28.000         NaN   
5      35152  2019-04-19 07:44:46.000  2019-04-19 07:44:46.000         NaN   
6      35149  2019-04-19 07:30:13.000  2019-04-19 07:30:13.000         NaN   
7      34660  2019-04-18 09:41:28.000  2019-04-18 09:41:28.000         NaN   
8      34663  2019-04-18 09:46:27.000  2019-04-18 09:46:27.000         NaN   
9      34951  2019-04-18 13:13:55.000  2019-04-18 13:13:55.000         NaN   

                    product_id_hash  stock_receipt_id  price   qty  \
0  ac6eed7114066ff5d9b10473b7ef6341              2404   0.72  3000   
1  2963dbc550404970787514aa177006e8              2401   5.70   558   
2  0b9dcc39ddecd3359566b59d3c9a6b9c              2410  31.81   100   
3  874121fee1d3bebea38815d169d47848              2407   0.72  3000   
4  6ab2ef9e3a9f4ed96fb1b77191e612d0              2401   3.60   319   
5  9c8001602e98208ef4e8d1bbf79fee65              2401   7.40   459   
6  4e18967c55baab1033250c9f8b0016b1              2404   0.72  3000   
7  9bf521d44fb17b1abc961ebc08c28de0              2371   5.02   300   
8  e46bc064f8e92ac2c404b9871b2a4ef2              2374  54.90    15   
9  b45232282ea62bfccffbd5350317e7e2              2383   3.60     8   

        status message  picker_id  
0     in_stock     NaN        NaN  
1     in_stock     NaN        NaN  
2  in_progress     NaN        NaN  
3     in_stock     NaN        NaN  
4     in_stock     NaN        NaN  
5     in_stock     NaN        NaN  
6     in_stock     NaN        NaN  
7     in_stock     NaN        NaN  
8     in_stock     NaN        NaN  
9  in_progress     NaN        NaN

From stock_products table I could use the following columns: *product_id_hash*, *qty*, *status*, *created_at*

In [5]:
sales_order = pd.read_csv('in.c_assignment_hash.sales_order.csv.gz', compression='gzip', sep=',')
sales_order.head(10)

entity_id    status                               shipping_description  \
0     165932  complete          Slovenská pošta - Doručenie na Vašu poštu   
1     927318  complete                            Doručenie na Vašu poštu   
2     904442  complete  Kikézbesítés GLS futárral - Kikézbesítés GLS f...   
3    1075000  canceled                             Doručenie kuriérom DHL   
4     593847  complete  Kikézbesítés GLS Csomagpont - Kikézbesítés GLS...   
5     647604  complete  Doručení na výdejnu Zásilkovna - Doručení na v...   
6     818759  complete                             Doručenie kuriérom DHL   
7     868947  complete                       Doručenie do parcelshopu DHL   
8     491844  complete  Kikézbesítés GLS Csomagpont - Kikézbesítés GLS...   
9     165495  complete                    GLS - Kikézbesítés GLS futárral   

   store_id  customer_id  base_subtotal base_currency_code  \
0       1.0          NaN          39.45                EUR   
1       1.0          NaN          17.42                EUR   
2       3.0     140932.0        3881.89                HUF   
3       1.0          NaN          46.50                EUR   
4       3.0          NaN        5346.46                HUF   
5       2.0      40315.0         380.87                CZK   
6       1.0      21875.0           8.50                EUR   
7       1.0          NaN          21.62                EUR   
8       3.0          NaN        3692.91                HUF   
9       NaN      34977.0           3.20                EUR   

   base_to_global_rate global_currency_code                    customer_email  \
0               1.0000                  EUR  40381b162ff8cdc7df99a28b60f35d46   
1               1.0000                  EUR  5c72af01665fe138a6a780ffde958410   
2               0.0032                  EUR  0177a29a445a6e049f594c3de2c4aad0   
3               1.0000                  EUR  6e072370366c669c51113485a2a200c8   
4               0.0032                  EUR  1edfdc4db492a0dade4fda787f937b7b   
5               0.0395                  EUR  e51d7064b080fe39595014c884911af8   
6               1.0000                  EUR  2b2be116adf3019d06616482b651b047   
7               1.0000                  EUR  b52d02e71c5923dfece68084db3946c9   
8               0.0032                  EUR  eadd673d1ecaf320bb5c8355415873c8   
9               1.0000                  EUR  9177408418b4cbfd669963bbe6d4ee50   

              shipping_method                          store_name  \
0           flatrate_flatrate     GymBeamSk\nGymBeamSk\nGymBeamSk   
1           slovakpost_pickup     GymBeamSk\nGymBeamSk\nGymBeamSk   
2       gls_delivery_delivery     GymBeamHu\nGymBeamHu\nGymBeamHu   
3       dhl_delivery_delivery     GymBeamSk\nGymBeamSk\nGymBeamSk   
4         gls_shipping_pickup     GymBeamHu\nGymBeamHu\nGymBeamHu   
5  zasilkovna_shipping_pickup     GymBeamCz\nGymBeamCz\nGymBeamCz   
6       dhl_delivery_delivery     GymBeamSk\nGymBeamSk\nGymBeamSk   
7         dhl_shipping_pickup     GymBeamSk\nGymBeamSk\nGymBeamSk   
8         gls_shipping_pickup     GymBeamHu\nGymBeamHu\nGymBeamHu   
9           flatrate_flatrate  GymBeamPnw\nGymBeamPnw\nGymBeamPnw   

            created_at           updated_at  customer_gender  
0  2016-11-19 17:16:25  2019-01-04 09:44:50              NaN  
1  2019-01-22 09:24:20  2019-01-30 09:34:13              NaN  
2  2019-01-09 18:30:12  2019-01-22 15:03:25              NaN  
3  2019-04-01 13:43:59  2019-04-01 15:33:00              NaN  
4  2018-05-08 17:14:50  2018-05-15 12:45:27              NaN  
5  2018-06-25 07:34:18  2018-07-03 13:45:37              0.0  
6  2018-11-16 19:29:02  2018-11-29 13:11:01              0.0  
7  2018-12-17 20:27:15  2019-01-04 16:45:20              NaN  
8  2018-02-06 21:41:50  2018-03-06 22:52:27              NaN  
9  2016-11-18 10:24:37  2019-01-04 09:44:50              NaN

Looks like the sales_order table does not contain any useful columns

In [6]:
sales_order_item = pd.read_csv('in.c_assignment_hash.sales_order_item.csv.gz', compression='gzip', sep=',')
sales_order_item.head(10)

item_id  order_id  store_id               created_at  \
0  3580877    659888       1.0  2018-07-07 13:48:40.000   
1  4620535    818022       3.0  2018-11-16 17:01:08.000   
2  3659222    672746       1.0  2018-07-17 17:46:13.000   
3  3099216    576715       1.0  2018-04-21 20:14:39.000   
4  3244677    602157       2.0  2018-05-15 10:33:36.000   
5  3715931    682068       1.0  2018-07-26 12:31:42.000   
6  5058712    883127       1.0  2019-01-02 09:23:48.000   
7   948026    200528       1.0  2017-09-06 03:50:43.000   
8   167932     42782       1.0  2017-09-05 16:21:06.000   
9  4011683    729785       3.0  2018-09-09 10:03:02.000   

                updated_at                   product_id_hash  product_type  \
0  2018-07-14 11:41:53.000  d187ef9ca106a882adda4068bf44ef9a  configurable   
1  2018-12-12 23:49:42.000  34f9a343f945196b66f807e0eb6249fd        simple   
2  2018-07-27 19:29:08.000  799fc232a527e7a7d6435c945083a329        simple   
3  2018-05-16 17:21:31.000  609a199881ca4ba9c95688235cd6ac5c        simple   
4  2018-05-23 14:41:02.000  5cf21ce30208cfffaa832c6e44bb567d  configurable   
5  2018-08-01 14:34:27.000  47257279d0b4f033e373b16e65f8f089        simple   
6  2019-01-14 05:19:42.000  e0854e3c03ec877be65d351b90680d46        simple   
7  2018-03-10 02:30:09.000  f9a1b3ab351b119e19ba3133c0a8fad5  configurable   
8  2018-03-10 02:12:04.000  fcac695db02687ffb7955b66a43fe6e6        simple   
9  2018-09-22 15:37:27.000  45c68484c6fc509cb25bdfca881e5cd8        simple   

                          name_hash  description  qty_ordered  base_price  \
0  b8b838e61e7ced13f70bea364738d061          NaN          2.0       14.75   
1  f98ec26b23b21e9464b6e45831d485de          NaN          1.0        0.00   
2  e154a2e173b8404006121dfd7716bfb5          NaN          1.0        0.00   
3  5a364afed14512eafee2723b8eb6d596          NaN          1.0        0.00   
4  56e9867f441183740e0fa21e6eddb585          NaN          2.0      242.61   
5  babb0b32eed886f7fccf7b1e1f6b38ce          NaN          1.0        0.00   
6  c6c213ffe4db7c93d75ef5649c704a08          NaN          1.0        0.00   
7  7a6bbcc03d201a1f786e3e82e81f679a          NaN          1.0        0.00   
8  7d1ba8eb8eabde89ec6487ea8ab90f3a          NaN          1.0        0.00   
9  1616e66a223684660fae754328b6682b          NaN          1.0        0.00   

   base_row_total  free_shipping  
0           29.50              0  
1            0.00              0  
2            0.00              0  
3            0.00              0  
4          485.22              0  
5            0.00              0  
6            0.00              0  
7            0.00              0  
8            0.00              0  
9            0.00              0

In [7]:
v_product_list = pd.read_csv('in.c_assignment_hash.v_product_list.csv.gz', compression='gzip', sep=',')
v_product_list.head(10)

website_id                    entity_id_hash       type_id  \
0           1  d67d8ab4f4c10bf22aa353e27879133c  configurable   
1           2  d67d8ab4f4c10bf22aa353e27879133c  configurable   
2           3  d67d8ab4f4c10bf22aa353e27879133c  configurable   
3           4  d67d8ab4f4c10bf22aa353e27879133c  configurable   
4           5  d67d8ab4f4c10bf22aa353e27879133c  configurable   
5           7  d67d8ab4f4c10bf22aa353e27879133c  configurable   
6           8  d67d8ab4f4c10bf22aa353e27879133c  configurable   
7          10  d67d8ab4f4c10bf22aa353e27879133c  configurable   
8          11  d67d8ab4f4c10bf22aa353e27879133c  configurable   
9           1  d645920e395fedad7bbbed0eca3fe2e0        simple   

                created_at                         name_hash  \
0  2017-09-04 20:24:21.000  2a4b04a2385be87059765e016b26323c   
1  2017-09-04 20:24:21.000  2a4b04a2385be87059765e016b26323c   
2  2017-09-04 20:24:21.000  b94e65268c8e918957351b534e1fc815   
3  2017-09-04 20:24:21.000  2a4b04a2385be87059765e016b26323c   
4  2017-09-04 20:24:21.000  2a4b04a2385be87059765e016b26323c   
5  2017-09-04 20:24:21.000  2a4b04a2385be87059765e016b26323c   
6  2017-09-04 20:24:21.000  2a4b04a2385be87059765e016b26323c   
7  2017-09-04 20:24:21.000  2a4b04a2385be87059765e016b26323c   
8  2017-09-04 20:24:21.000  2a4b04a2385be87059765e016b26323c   
9  2017-09-04 20:24:24.000  3c76748d681e90c0add26ca059cf971a   

                  default_name_hash  cost    price  special_price  \
0  2a4b04a2385be87059765e016b26323c   NaN     58.0            NaN   
1  2a4b04a2385be87059765e016b26323c   NaN   1530.0            NaN   
2  2a4b04a2385be87059765e016b26323c   NaN  20870.0            NaN   
3  2a4b04a2385be87059765e016b26323c   NaN    252.9            NaN   
4  2a4b04a2385be87059765e016b26323c   NaN     58.0            NaN   
5  2a4b04a2385be87059765e016b26323c   NaN     58.0            NaN   
6  2a4b04a2385be87059765e016b26323c   NaN     58.0            NaN   
7  2a4b04a2385be87059765e016b26323c   NaN     58.0            NaN   
8  2a4b04a2385be87059765e016b26323c   NaN     58.0            NaN   
9  3c76748d681e90c0add26ca059cf971a   NaN     58.0           55.0   

   manufacturer  is_in_stock  status  parent_id  long_term_unavaible  
0          13.0            0       1        NaN                  1.0  
1          13.0            0       1        NaN                  1.0  
2          13.0            0       1        NaN                  1.0  
3          13.0            0       1        NaN                  1.0  
4          13.0            0       1        NaN                  1.0  
5          13.0            0       1        NaN                  1.0  
6          13.0            0       1        NaN                  1.0  
7          13.0            0       1        NaN                  1.0  
8          13.0            0       1        NaN                  1.0  
9           0.0            0       1       39.0                  1.0

From v_product_list table I could use the following columns: *entity_id_hash*, *type_id*, *is_in_stock*, *long_term_unavailable*

First let's have a look at the product list table where both long_term_unavailable and type_id (configurable or simple product) are listed and this data is needed for further analysis

In [8]:
# check how many rows & columns are in the table
v_product_list.shape

(54809, 14)

There are 54809 rowns and 14 columns in the v_product_list table

In [9]:
# check the dataypes of columns
v_product_list.dtypes

website_id               int64
entity_id_hash          object
type_id                 object
created_at              object
name_hash               object
default_name_hash       object
cost                   float64
price                  float64
special_price          float64
manufacturer           float64
is_in_stock              int64
status                   int64
parent_id              float64
long_term_unavaible    float64
dtype: object

In [10]:
# check count of missing values
v_product_list.isna().sum()

website_id                 0
entity_id_hash             0
type_id                    0
created_at                 0
name_hash                  0
default_name_hash          0
cost                    4454
price                   7590
special_price          12285
manufacturer             526
is_in_stock                0
status                     0
parent_id              12570
long_term_unavaible      543
dtype: int64

543 rows of information about long_term_available products is missing. Since I do not have information to be able to add the data and I only need information for available products I will ignore the missing data.

In [11]:
# check if and how many duplicate rows are there in the table
v_product_list.duplicated().sum()

92

There are 92 duplicate rows in the table so I'll remove them from the table

In [12]:
# delete duplicated rows
v_product_list.drop_duplicates(inplace = True)

In [13]:
# check the count of unique encrypted product ID
v_product_list.entity_id_hash.nunique()

8793

There are 8793 unique product IDs in the v_product_list table

Let's have a closer look at categorical and numerical values of the v_product_list table

In [14]:
# check the categorical values
v_product_list.describe(include = 'object')

entity_id_hash type_id               created_at  \
count                              54717   54717                    54717   
unique                              8793       4                     8783   
top     c74214a3877c4d8297ac96217d5189b7  simple  2017-09-05 08:33:54.000   
freq                                  45   42943                       45   

                               name_hash                 default_name_hash  
count                              54717                             54717  
unique                             13504                              8428  
top     5704a405db8e6228c063f2bc8b68179c  91f1895d5c4b34791efc4147587425e7  
freq                                  46                                45

In [15]:
# check the numerical vaules
v_product_list.describe()

website_id          cost         price  special_price  manufacturer  \
count  54717.000000  50303.000000  47154.000000   42485.000000  54191.000000   
mean       5.152000     10.136746   1338.281260     999.556392    163.673064   
std        3.702283     10.658311   3873.351986    3127.159436    385.405634   
min        1.000000      0.000000      0.000000       0.000000      0.000000   
25%        2.000000      2.920000     14.950000      12.000000     32.000000   
50%        4.000000      6.500000     47.600000      37.900000     44.000000   
75%        8.000000     13.400000    359.000000     261.600000     64.000000   
max       13.000000    110.000000  71960.000000  127000.000000   1558.000000   

        is_in_stock        status     parent_id  long_term_unavaible  
count  54717.000000  54717.000000  42200.000000         54174.000000  
mean       0.410384      1.013433  11956.032536             0.210415  
std        0.491908      0.115120  10427.740551             0.407607  
min        0.000000      1.000000     39.000000             0.000000  
25%        0.000000      1.000000   4960.250000             0.000000  
50%        0.000000      1.000000   7306.000000             0.000000  
75%        1.000000      1.000000  28140.000000             0.000000  
max        1.000000      2.000000  30922.000000             1.000000

In [16]:
# create table for only simple and long term available products and display random 10 rows
v_product_list_available_simple = v_product_list.loc[(v_product_list['type_id'] == 'simple') & (v_product_list['long_term_unavaible'] == 0.0)]
v_product_list_available_simple.sample(10)

website_id                    entity_id_hash type_id  \
28676          11  ac8a9143597891fc2fc2ded41a9a9ec7  simple   
22793           5  0f7b2bb5d0c7e6209ecff113108f64cf  simple   
19463           2  198dd5fb9c43b2d29a548f8c77e85cf9  simple   
16084          12  fc95fa5740ba01a870cfa52f671fe1e4  simple   
43697          11  e0e93ef9fba9598319005e69e0d48499  simple   
26781           1  ac6b3cce8c74b2e23688c3e45532e2a7  simple   
51888           8  a43cf1bfb47754486720d964df63425b  simple   
38527           2  7c88b74f49d5cfac39702e8e947cb0c0  simple   
46304           8  5262d5a886484791a7c4c9bd5351a2cf  simple   
53377          12  24d26100b9d2097f0418ed14dd528b38  simple   

                    created_at                         name_hash  \
28676  2017-09-05 06:55:05.000  80061e59c256bbf25f0b07d085313207   
22793  2017-09-05 05:30:25.000  d107e081c7f0c2ebf32d767dd6475f4c   
19463  2017-09-05 01:09:11.000  cfa01d7442caf9cbb7d0ba6026e1c7b1   
16084  2017-09-05 00:36:37.000  128d220a63e7689be77abf02ca5cb519   
43697  2018-05-07 11:04:44.000  bbb532bff815c5a3d78417107087e4d6   
26781  2017-09-05 06:27:40.000  60d1ae8c75a2ad910bcaaccde3ee1e77   
51888  2019-01-25 13:11:30.000  068f96315415af84d6e23f656365562c   
38527  2017-10-06 14:14:26.000  df4b8043090bd44ea8af36db62c1878b   
46304  2018-07-12 11:50:43.000  77d61a546a554fc71611aa863682f284   
53377  2019-02-20 08:02:09.000  edca2a482c112e2e1ba4710816664041   

                      default_name_hash     cost   price  special_price  \
28676  80061e59c256bbf25f0b07d085313207  14.1679  176.00         145.00   
22793  d107e081c7f0c2ebf32d767dd6475f4c   2.5900   14.95           9.95   
19463  cfa01d7442caf9cbb7d0ba6026e1c7b1   6.2000  527.00         444.00   
16084  128d220a63e7689be77abf02ca5cb519  27.4398   43.95          41.20   
43697  bbb532bff815c5a3d78417107087e4d6   3.0900  119.00          85.00   
26781  60d1ae8c75a2ad910bcaaccde3ee1e77   0.6200    2.95           2.30   
51888  068f96315415af84d6e23f656365562c  12.9500   29.95          24.95   
38527  df4b8043090bd44ea8af36db62c1878b   2.7400  416.00         279.00   
46304  77d61a546a554fc71611aa863682f284  16.0000   29.49          23.49   
53377  edca2a482c112e2e1ba4710816664041   3.0200   14.95           9.95   

       manufacturer  is_in_stock  status  parent_id  long_term_unavaible  
28676           9.0            1       1     7960.0                  0.0  
22793          64.0            0       1     7047.0                  0.0  
19463          32.0            1       1     6454.0                  0.0  
16084          33.0            1       1     5889.0                  0.0  
43697          64.0            0       1    28893.0                  0.0  
26781         118.0            0       1     7693.0                  0.0  
51888        1482.0            1       1    30052.0                  0.0  
38527          64.0            0       1    28206.0                  0.0  
46304          40.0            0       1     4948.0                  0.0  
53377          64.0            1       1    30257.0                  0.0

Let's display only columns wchich can be beneficial for my analysis

In [17]:
# display only simple unique encrypted product IDs which are available
v_product_list_available_simple_basic = v_product_list_available_simple [['entity_id_hash', 'type_id', 'is_in_stock', 'long_term_unavaible']]
v_product_list_available_simple_basic

entity_id_hash type_id  is_in_stock  \
39     072b030ba126b2f4b2374f342be9ed44  simple            1   
40     072b030ba126b2f4b2374f342be9ed44  simple            1   
41     072b030ba126b2f4b2374f342be9ed44  simple            1   
42     072b030ba126b2f4b2374f342be9ed44  simple            1   
43     072b030ba126b2f4b2374f342be9ed44  simple            1   
...                                 ...     ...          ...   
54804  874121fee1d3bebea38815d169d47848  simple            1   
54805  874121fee1d3bebea38815d169d47848  simple            1   
54806  874121fee1d3bebea38815d169d47848  simple            1   
54807  874121fee1d3bebea38815d169d47848  simple            1   
54808  874121fee1d3bebea38815d169d47848  simple            1   

       long_term_unavaible  
39                     0.0  
40                     0.0  
41                     0.0  
42                     0.0  
43                     0.0  
...                    ...  
54804                  0.0  
54805                  0.0  
54806                  0.0  
54807                  0.0  
54808                  0.0  

[35169 rows x 4 columns]

From the above table it's obvious that there are duplicate product IDs - let's check how many unique ID's are there in the v_product_list table which also are in stock 

In [18]:
v_product_list_available_simple_basic.loc[v_product_list_available_simple_basic['is_in_stock'] == 1].nunique()

entity_id_hash         1848
type_id                   1
is_in_stock               1
long_term_unavaible       1
dtype: int64

There are 1848 unique available simple products which are in stock according to v_product_list table

In [19]:
# display table of available simple products which are in stock
v_product_list_available_simple_in_stock_basic = v_product_list_available_simple_basic.loc[v_product_list_available_simple_basic['is_in_stock'] == 1]
v_product_list_available_simple_in_stock_basic

entity_id_hash type_id  is_in_stock  \
39     072b030ba126b2f4b2374f342be9ed44  simple            1   
40     072b030ba126b2f4b2374f342be9ed44  simple            1   
41     072b030ba126b2f4b2374f342be9ed44  simple            1   
42     072b030ba126b2f4b2374f342be9ed44  simple            1   
43     072b030ba126b2f4b2374f342be9ed44  simple            1   
...                                 ...     ...          ...   
54804  874121fee1d3bebea38815d169d47848  simple            1   
54805  874121fee1d3bebea38815d169d47848  simple            1   
54806  874121fee1d3bebea38815d169d47848  simple            1   
54807  874121fee1d3bebea38815d169d47848  simple            1   
54808  874121fee1d3bebea38815d169d47848  simple            1   

       long_term_unavaible  
39                     0.0  
40                     0.0  
41                     0.0  
42                     0.0  
43                     0.0  
...                    ...  
54804                  0.0  
54805                  0.0  
54806                  0.0  
54807                  0.0  
54808                  0.0  

[14949 rows x 4 columns]

Let's have a look at the stock_days_data table to determine how many products based on product ID are there in stock according to this database and how the quantity changes on various dates - as here is also the 'quantity' column which we will use in further analysis

In [20]:
# check how many rows and columns the stock_products table has
stock_days_data.shape

(7622765, 6)

The stock_products table has 7622765 rows and 6 columns

In [21]:
# review data types
stock_days_data.dtypes

id                   int64
period              object
product_id_hash     object
is_in_stock          int64
qty                float64
cost               float64
dtype: object

In [22]:
# check missing values
stock_days_data.isna().sum()

id                    0
period                0
product_id_hash       0
is_in_stock           0
qty                7837
cost                998
dtype: int64

7837 quantity entries are missing

In [23]:
# check if there are any duplicate rows
stock_days_data.duplicated().sum()

0

There are no duplicate rows in the stock_days_data table

In [24]:
# check number of unique product IDs in stock
stock_days_data.product_id_hash.nunique()

13116

There are 13116 unique product IDs in the stock_products table

Only product IDs which have 'in_stock' status are to be considered for next analysis

In [25]:
stock_days_data.loc[stock_days_data['is_in_stock'] == 1]

id      period                   product_id_hash  is_in_stock  \
1        9992268  2019-02-23  7ca595cd7955654b36629c1470763487            1   
3        9986736  2019-02-23  4ea83d951990d8bf07a68ec3e50f9156            1   
10       9992278  2019-02-23  923afdc34efd0d626a7d62c49f3fb4e8            1   
14       9990173  2019-02-23  076ccd93ad68be51f23707988e934906            1   
18       9991951  2019-02-23  20a8571b66205bd36a898172ae082c53            1   
...          ...         ...                               ...          ...   
7622746  4786803  2017-10-25  eaae339c4d89fc102edd9dbdb6a28915            1   
7622748  5813130  2017-12-14  02ed812220b0705fabb868ddbf17ea20            1   
7622752  1987520  2017-09-27  c1aeb6517a1c7f33514f7ff69047e74e            1   
7622755  2868196  2017-10-10  cc6ef8cb8df3af5693726838cd728163            1   
7622758  4317692  2017-10-21  d27b95cac4c27feb850aaa4070cc4675            1   

           qty      cost  
1        297.0    1.2744  
3         56.0   19.6600  
10        26.0    5.3100  
14         4.0   43.5100  
18        10.0   12.0000  
...        ...       ...  
7622746  935.0    2.6700  
7622748   20.0    4.3100  
7622752    1.0   78.8100  
7622755   76.0  230.5800  
7622758    7.0    8.2900  

[2363759 rows x 6 columns]

There are 2363759 products which are in stock

Let's have a look at stock_days_data summaries

In [26]:
# review categorical data summary
stock_days_data.describe(include = 'all')

id      period                   product_id_hash  \
count   7.622765e+06     7622765                           7622765   
unique           NaN         603                             13116   
top              NaN  2017-10-28  19e901474bd32d47931f0219992ff889   
freq             NaN       66079                              1212   
mean    5.071095e+06         NaN                               NaN   
std     3.276491e+06         NaN                               NaN   
min     1.000000e+00         NaN                               NaN   
25%     2.003686e+06         NaN                               NaN   
50%     4.956398e+06         NaN                               NaN   
75%     8.134678e+06         NaN                               NaN   
max     1.116933e+07         NaN                               NaN   

         is_in_stock           qty          cost  
count   7.622765e+06  7.614928e+06  7.621767e+06  
unique           NaN           NaN           NaN  
top              NaN           NaN           NaN  
freq             NaN           NaN           NaN  
mean    3.151164e-01  3.516597e+01  2.857840e+02  
std     1.536153e+00  3.245811e+02  1.524882e+03  
min     0.000000e+00 -2.420000e+02 -6.820000e-02  
25%     0.000000e+00  0.000000e+00  5.500000e+00  
50%     0.000000e+00  0.000000e+00  1.197000e+01  
75%     1.000000e+00  5.000000e+00  2.818000e+01  
max     5.500000e+02  8.631800e+04  1.000000e+05

Let's take into consideration only 'in_stock' products for my further data analysis

In [27]:
# create new table with products with only 'in-stock' status and display 10 random rows  
stock_days_data_in_stock = stock_days_data.loc[stock_days_data['is_in_stock'] == 1]
stock_days_data_in_stock.sample(10)

id      period                   product_id_hash  is_in_stock  \
5482640   8848618  2018-10-17  57b9c682ed39822cdebb3c80d823794b            1   
948351    7304406  2018-04-14  109f91266ef89cc3690079b28abfe9a3            1   
5192983  10036841  2019-03-02  0853de1f652d3a206e18264d18eb3bb0            1   
4719242   1548445  2017-09-20  7f687767ccf20fcea1c9dc4a5adc2326            1   
2691303   7136865  2018-03-25  2a8a8bde56a1a353f4e5fdd641f0b199            1   
7492870   2405347  2017-10-03  6f5e4e86a87220e5d361ad82f1ebc335            1   
6784449   8578301  2018-09-15  edf0320adc8658b25ca26be5351b6c4a            1   
7306251   2393159  2017-10-03  1bc2029a8851ad344a8d503930dfd7f7            1   
2522527   3524388  2017-10-15  786f2d0acade1787dacb21cac85d099c            1   
7334757    213131  2017-08-16  2afdb805c17991d2cdeb3b51b1d77857            1   

           qty   cost  
5482640   42.0   2.92  
948351    18.0  14.57  
5192983   28.0   9.00  
4719242    1.0  34.77  
2691303    4.0   5.60  
7492870   49.0   8.29  
6784449    6.0   1.76  
7306251   95.0   1.69  
2522527  114.0   6.25  
7334757    5.0   7.47

In [28]:
# display only necessary columns - period, product_id_hash and quantity
stock_days_data_in_stock_basic = stock_days_data_in_stock[['period', 'product_id_hash', 'qty']]
stock_days_data_in_stock_basic.sample(10)

period                   product_id_hash    qty
1601399  2018-05-08  5a0ff4520c5b59aae5315322f3927d39    3.0
3967533  2017-10-11  358f9e7be09177c17d0d17ff73584307   30.0
2848871  2017-08-21  ab38039d7c1a6862cc692cd15e20d757  955.0
424846   2017-09-12  95cc7ef498e141173576365264fc5fba    2.0
374206   2018-02-23  46dce5f2f0e61edb70931a00d00a464e   31.0
525341   2017-09-14  cf5530d9e441e0d78574353214373569   13.0
5871934  2019-01-31  ad5ab36761669d6eadbaee691c4a1d22    9.0
6243683  2018-05-30  75da5036f659fe64b53f3d9b39412967    4.0
4200602  2017-09-08  1c63926ebcabda26b5cdb31b5cc91efb    9.0
6459600  2017-10-29  7e185cc0ad0a719c730af5354d7142c1    7.0

Let's double check if there aren't any missing values and how many unique product IDs are there in stock

In [29]:
# display number of missing values
stock_days_data_in_stock_basic.isna().sum()

period             0
product_id_hash    0
qty                0
dtype: int64

In [30]:
# display how many unique product IDs are there
stock_days_data_in_stock_basic.product_id_hash.nunique()

6330

Now let's merge the data frames v_product_list_available_simple_in_stock_basic and stock_days_data_in_stock_basic.
Since I only need data for simple and available products, I will need to have displayed each row from the v_product_list_available_simple_basic table and matching number of rows of the stock_days_data_in_stock_basic table

In [31]:
# create table using inner join
available_simple_products_in_stock = pd.merge(v_product_list_available_simple_in_stock_basic, stock_days_data_in_stock_basic, left_on = 'entity_id_hash', right_on = 'product_id_hash', how = 'inner')
available_simple_products_in_stock

entity_id_hash type_id  is_in_stock  \
0        072b030ba126b2f4b2374f342be9ed44  simple            1   
1        072b030ba126b2f4b2374f342be9ed44  simple            1   
2        072b030ba126b2f4b2374f342be9ed44  simple            1   
3        072b030ba126b2f4b2374f342be9ed44  simple            1   
4        072b030ba126b2f4b2374f342be9ed44  simple            1   
...                                   ...     ...          ...   
8566946  874121fee1d3bebea38815d169d47848  simple            1   
8566947  874121fee1d3bebea38815d169d47848  simple            1   
8566948  874121fee1d3bebea38815d169d47848  simple            1   
8566949  874121fee1d3bebea38815d169d47848  simple            1   
8566950  874121fee1d3bebea38815d169d47848  simple            1   

         long_term_unavaible      period                   product_id_hash  \
0                        0.0  2019-02-25  072b030ba126b2f4b2374f342be9ed44   
1                        0.0  2019-02-27  072b030ba126b2f4b2374f342be9ed44   
2                        0.0  2019-02-26  072b030ba126b2f4b2374f342be9ed44   
3                        0.0  2019-02-28  072b030ba126b2f4b2374f342be9ed44   
4                        0.0  2018-12-28  072b030ba126b2f4b2374f342be9ed44   
...                      ...         ...                               ...   
8566946                  0.0  2019-04-22  874121fee1d3bebea38815d169d47848   
8566947                  0.0  2019-04-19  874121fee1d3bebea38815d169d47848   
8566948                  0.0  2019-04-20  874121fee1d3bebea38815d169d47848   
8566949                  0.0  2019-04-21  874121fee1d3bebea38815d169d47848   
8566950                  0.0  2019-04-22  874121fee1d3bebea38815d169d47848   

            qty  
0          75.0  
1          55.0  
2          63.0  
3          48.0  
4         543.0  
...         ...  
8566946  2994.0  
8566947  3000.0  
8566948  3000.0  
8566949  2999.0  
8566950  2994.0  

[8566951 rows x 7 columns]

In [32]:
# double check if there are any missing values
available_simple_products_in_stock.isna().sum()

entity_id_hash         0
type_id                0
is_in_stock            0
long_term_unavaible    0
period                 0
product_id_hash        0
qty                    0
dtype: int64

There are no missing values in the merged table

In [33]:
# check how many unique values are there
available_simple_products_in_stock.nunique()

entity_id_hash         1848
type_id                   1
is_in_stock               1
long_term_unavaible       1
period                  596
product_id_hash        1848
qty                    5048
dtype: int64

Display the table with necessary columns only

In [34]:
available_simple_products_in_stock_simple = available_simple_products_in_stock[['entity_id_hash', 'period', 'qty']]
available_simple_products_in_stock_simple

entity_id_hash      period     qty
0        072b030ba126b2f4b2374f342be9ed44  2019-02-25    75.0
1        072b030ba126b2f4b2374f342be9ed44  2019-02-27    55.0
2        072b030ba126b2f4b2374f342be9ed44  2019-02-26    63.0
3        072b030ba126b2f4b2374f342be9ed44  2019-02-28    48.0
4        072b030ba126b2f4b2374f342be9ed44  2018-12-28   543.0
...                                   ...         ...     ...
8566946  874121fee1d3bebea38815d169d47848  2019-04-22  2994.0
8566947  874121fee1d3bebea38815d169d47848  2019-04-19  3000.0
8566948  874121fee1d3bebea38815d169d47848  2019-04-20  3000.0
8566949  874121fee1d3bebea38815d169d47848  2019-04-21  2999.0
8566950  874121fee1d3bebea38815d169d47848  2019-04-22  2994.0

[8566951 rows x 3 columns]

In [35]:
# check datatypes to determine if any need to be amended
available_simple_products_in_stock_simple.dtypes

entity_id_hash     object
period             object
qty               float64
dtype: object

Period column needs to be converted to datetime to be able count the days so let's import the respective library

In [36]:
import datetime as dt

In [37]:
# convert object data type to datetime type
pd.to_datetime(available_simple_products_in_stock_simple['period'])

0         2019-02-25
1         2019-02-27
2         2019-02-26
3         2019-02-28
4         2018-12-28
             ...    
8566946   2019-04-22
8566947   2019-04-19
8566948   2019-04-20
8566949   2019-04-21
8566950   2019-04-22
Name: period, Length: 8566951, dtype: datetime64[ns]

Final table with product IDs and average quantity of the product in stock as per date

In [43]:
Series = available_simple_products_in_stock_simple.groupby(['entity_id_hash', 'period'])['qty'].mean()
Final = Series.to_frame('avg quantity')
Final.head(100)


avg quantity
entity_id_hash                   period                  
001ab2fa029c064a45e41f8b2644a292 2017-09-03          13.0
                                 2017-09-05          12.0
                                 2017-09-06          12.0
                                 2017-09-07          11.0
                                 2017-09-08          11.0
...                                                   ...
                                 2019-03-06           3.0
                                 2019-03-19          10.0
                                 2019-03-20           8.0
                                 2019-03-21           6.0
                                 2019-03-22           6.0

[100 rows x 1 columns]